pkg/state/state.go

```diff
@@ -87,25 +86,27 @@ type ReleaseSetSpec struct {
 // helmStateAlias is helm state alias
 type helmStateAlias HelmState
 
-func (hs HelmState) MarshalYAML() (interface{}, error) {
-	return helmStateAlias(hs), nil
+func (hs HelmState) MarshalYAML() ([]byte, error) {
+	hsa := helmStateAlias(hs)
+	return yaml.Marshal(hsa)
 }
 
-func (hs *HelmState) UnmarshalYAML(value *yaml.Node) error {
```

Comment:
@yxxhero I believe you can revert 589576e and make it back to use `UnmarshalYaml(func(interface{}) error) error`, as that's what we've been implementing when we were originally using gopkg.in/yaml.v2 alone and goccy/go-yaml does support the exact interface too https://github.com/goccy/go-yaml/blob/6cdefc42e112ac71cbe316e1eed264ea62f58e25/yaml.go#L49-L52

Reply:
HelmState's UnmarshalYAML was created for v1, and I feel like it can be adapted to the latest goccy/go-yaml interface, which I don't need to introduce for v0. @mumoshu

Reply:
I've updated my comment before you replied! Sorry for the confusion. Regardless of when it was introduced first, I think the problem of using the new interface seems to have been caught by the test failure- it now does not intend nested YAML dicts correctly.

Reply:
yes. should we post an issue to goccy/go-yaml?

Reply:
Why? We don't need to stick with `MarshalYAML() ([]byte, error) {`  nor `UnmarshalYAML(b []byte) error`, right?
Or do you know anything that isn't covered by our current test suite?

Reply:
Note that all the tests were passing at [b2515a7](https://github.com/helmfile/helmfile/pull/604/commits/b2515a7a6d48601a59bf76faf05965ff91c0b770)

Reply:
@yxxhero Maybe you wanted to keep using the `UnmarshalYAML(b []byte) error` assuming assuming it would be needed for the strict parsing even for goccy/go-yaml, right?

Reply:
Apparently goccy's respect strict parsing setting even you used InterfaceUnmarshaller to unmarshal nested structs. See https://github.com/goccy/go-yaml/blob/6cdefc42e112ac71cbe316e1eed264ea62f58e25/decode.go#L626 and https://github.com/goccy/go-yaml/blob/6cdefc42e112ac71cbe316e1eed264ea62f58e25/decode.go#L1045-L1050.

It would be great if we could add a new test case to test the desired behavior though... Any idea about that?

Reply:
yeah. I think so. when use `UnmarshalYAML(b []byte) error` for `HelmState`. we need a strict mode to keep the same behavior with the `yaml.UnMarshal`.

Reply:
@mumoshu

Reply:
@yxxhero I've reverted UnmarshalYAML and added a new test to ensure that it does respect the UnknownField option. The test is passing. WDYT?
https://github.com/helmfile/helmfile/pull/604/commits/d323cc1841997d0467404cead63de6ecf5c56e56

Reply:
good work.